lib/membrane_ffmpeg_video_filter/text_overlay.ex

```diff
@@ -90,50 +123,98 @@ defmodule Membrane.FFmpeg.VideoFilter.TextOverlay do
   end
 
   @impl true
-  def handle_caps(
-        :input,
-        %Raw{format: format, width: width, height: height} = caps,
-        _context,
-        state
-      ) do
+  def handle_caps(:input, caps, _context, state) do
+    state = init_new_filter_if_needed(caps, state)
+    {{:ok, caps: {:output, caps}, redemand: :output}, state}
+  end
+
+  @impl true
+  def handle_process(:input, buffer, ctx, state) do
+    {buffer, state} = apply_filter_if_needed(buffer, ctx, state)
+    {{:ok, [buffer: {:output, buffer}]}, state}
+  end
+
+  # no text left to render
+  defp apply_filter_if_needed(buffer, _ctx, %{text_stream: []} = state) do
+    {buffer, state}
+  end
+
+  defp apply_filter_if_needed(
+         %{metadata: metadata, payload: payload} = buffer,
+         ctx,
+         %{native_state: native_state, text_stream: [{interval, _text} | streams]} = state
+       ) do
+    cond do
+      frame_before_interval?(metadata, interval) ->
+        {buffer, state}
+
+      frame_after_interval?(metadata, interval) ->
+        state = %{state | text_stream: streams}
+        state = init_new_filter_if_needed(ctx.pads.input.caps, state)
+        apply_filter_if_needed(buffer, ctx, state)
+
+      frame_in_interval?(metadata, interval) ->
+        case Native.apply_filter(payload, native_state) do
+          {:ok, frame} ->
+            {%{buffer | payload: frame}, state}
+
+          {:error, reason} ->
+            raise inspect(reason)
```

Comment:
This can be moved to `Native` module - add an Elixir wrapper called `apply_filter!` raising on error